### PR TITLE
EOS-26878: Validation check for component and services.

### PR DESCRIPTION
**Pull request by shriya-deshmukh**
Signed-off-by: Shriya Deshmukh <shriya.deshmukh@seagate.com>

# Problem Statement
- Validation checks to ensure each component and service defined in cluster.yaml is supported in Utils const.py file.

# Design
-  Read component from cluster.yaml and verify it is present in const.py file.
- Read services list for component and verify service name is present in const.py file for that specific component.

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: src/provisioner/validators.py [hunk 1]
```diff
@@ -148,29 +147,60 @@ def _check_number_of_nodes(self):
...
+        for service in services:
+            # check if service name is define in utils const.py file.
+            if service not in Const._value2member_map_:
+                raise CortxProvisionerError(errno.EINVAL,
+                    f'{service} is not present in constant file.')
+            # Verify service_name is define for the specific component.
+            # Get all keys from constant file which has same 'service_name'.
+            constant_service_keys = [
+                key for key, enum_ele in Const.__members__.items() if enum_ele.value == service]
+            if not any(component_name.upper() in key for key in constant_service_keys):
```
Comment (ujjwalpl) [inline src/provisioner/validators.py:197]: service names are case sensitive. Comparison should also be.

File: src/provisioner/validators.py [hunk 2]
```diff
@@ -148,29 +147,60 @@ def _check_number_of_nodes(self):
...
+                    cortx_config_ext_services[service]['endpoints']):
                 raise CortxProvisionerError(errno.EINVAL,
                     f'{service} endpoints define in conf_store is not same as '
-                    'solution_config.')
+                    'cortx config.')
+        return 0
+
+    def _validate_components(self):
+        """Verify components defined in cluster.yaml is supported in constant file."""
+        node_types = self._get_config('cluster>node_types')
```
Comment (ujjwalpl) [inline src/provisioner/validators.py:173]: This one will not work with consul based conf store. Need to get exact key.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/provisioner/validators.py (hunk 2 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py:173]: This key I am reading from cluster.yaml file and not from the confstore.

File: src/provisioner/validators.py (hunk 1 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py:197]: Done.

**Review by ujjwalpl — APPROVED** [review]

File: src/provisioner/validators.py [hunk 3]
```diff
@@ -148,30 +147,61 @@ def _check_number_of_nodes(self):
...
+                    'cortx config.')
         return 0
 
-    def _get_value_from_solution_config(self, key):
-        """ Read config value for key from solution_config """
+    def _validate_components(self):
+        """Verify components defined in cluster.yaml is supported in constant file."""
+        node_types = self._get_config('cluster>node_types')
+        for node_type in node_types:
+            components_list = node_type['components']
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:175]: `component_list`

File: src/provisioner/validators.py [hunk 4]
```diff
@@ -148,30 +147,61 @@ def _check_number_of_nodes(self):
...
+            for component in components_list:
+                # check component_name exist in Utils Const cls.
+                component_name = component['name']
+                if component_name not in Const._value2member_map_:
+                    raise CortxProvisionerError(errno.EINVAL,
+                        f'{component_name} is not present in constant file.')
+                if component.get('services') is not None:
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:182]: Lets get value only once like:
service_list = component.get('services')

File: src/provisioner/validators.py [hunk 5]
```diff
@@ -148,30 +147,61 @@ def _check_number_of_nodes(self):
...
+                    services = component.get('services')
+                    self._validate_services(services, component_name)
+
+    def _validate_services(self, services, component_name):
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:186]: def _validate_services(self, service_list, component_name)

File: src/provisioner/validators.py [hunk 6]
```diff
@@ -148,30 +147,61 @@ def _check_number_of_nodes(self):
...
+        """Verify services defined in cluster.yaml is supported in constant file."""
+        for service in services:
+            # check if service name is define in utils const.py file.
+            if service not in Const._value2member_map_:
+                raise CortxProvisionerError(errno.EINVAL,
+                    f'{service} is not present in constant file.')
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:192]: Make error generic something like:
Value not supported...

File: src/provisioner/validators.py [hunk 7]
```diff
@@ -148,30 +147,61 @@ def _check_number_of_nodes(self):
...
+            # Verify service_name is define for the specific component.
+            # Get all keys from constant file which has same 'service_name'.
+            constant_service_keys = [
+                key for key, enum_ele in Const.__members__.items() if enum_ele.value.lower() == service.lower()]
+            if not any(component_name.upper() in key for key in constant_service_keys):
+                raise CortxProvisionerError(errno.EINVAL,
+                    f'"{service}" service defined in "{self.solution_conf_url}" for '
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:199]: Make error generic something like:
Component does not support service

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: src/provisioner/validators.py [hunk 8]
```diff
@@ -90,37 +91,35 @@ def load_config(cls, solution_conf_url, cortx_conf_url):
         cls.solution_conf_url = solution_conf_url
         cls.cortx_conf_url = cortx_conf_url
         Conf.load(cls._solution_index, cls.solution_conf_url)
-        cls.cortx_config_store = ConfigStore(cls.cortx_conf_url)
+        cls.cortx_conf = ConfigStore(cls.cortx_conf_url)
 
     def _check_storage_sets(self):
         """ Check no of storage_sets present in conf_store
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: done.

File: src/provisioner/validators.py [hunk 9]
```diff
@@ -90,37 +91,35 @@ def load_config(cls, solution_conf_url, cortx_conf_url):
...
-                f'is {len(input_storage_sets)} and in {self.cortx_conf_url} '
-                f'is {len(output_storage_sets)}')
+                f'cluster config. Storage_sets in {self.solution_conf_url} '
+                f'is {len(cluster_config_storage_sets)} and in {self.cortx_conf_url} '
+                f'is {len(confstore_storage_sets)}')
 
         return 0
 
     def _check_number_of_nodes(self):
         """ Check no. of nodes present in conf_store
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: Done.

File: src/provisioner/validators.py (hunk 3 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: Done.

File: src/provisioner/validators.py (hunk 4 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: Done.

File: src/provisioner/validators.py (hunk 5 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: Done

File: src/provisioner/validators.py (hunk 7 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: Done

File: src/provisioner/validators.py (hunk 6 above)
Comment (shriya-deshmukh) [reply src/provisioner/validators.py]: Done

File: src/provisioner/validators.py [hunk 10]
```diff
@@ -90,37 +92,32 @@ def load_config(cls, solution_conf_url, cortx_conf_url):
...
     def _check_storage_sets(self):
-        """ Check no of storage_sets present in conf_store
-            and solution_config is same. """
-
-        input_storage_sets = self._get_value_from_solution_config(
+        """Validate storage_sets present in cortx_conf."""
+        cluster_config_storage_sets = self._get_config(
             ConfigValidator._key_storage_set_sc)
-        output_storage_sets = self._get_value_from_conf_store(
+        confstore_storage_sets = self._get_value_from_conf_store(
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:101]: Change this:
`conf_store_storage_sets`

File: src/provisioner/validators.py [hunk 11]
```diff
@@ -133,45 +130,77 @@ def _check_number_of_nodes(self):
...
+                    f' Nodes in {self.solution_conf_url} is {len(cluster_config_nodes)} and '
+                    f'in {self.cortx_conf_url} is {len(conf_store_nodes)}')
 
         return 0
 
     def _check_external_services(self):
         """" Validate no of external services in cortx config """
 
-        sol_conf_ext_services = self._get_value_from_solution_config(
+        config_ext_services = self._get_config(
```
Comment (nitin-seagate) [inline src/provisioner/validators.py:148]: Change this to:
`solution_config_ext_services`

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

**Review by nitin-seagate — APPROVED** [review]
LGTM

**Pull request merged**